### DISCUSSION TITLE
Extrapolate from gauss points to nodes

### DISCUSSION THREAD

**Discussion by KhaledNabilSharafeldin** [idaholab/moose#21701] [Q&A General] [answered]
Hello,
Just a simple question here, I have in the main app the stress calculated at the gauss points, and whenever I transferred them to the sub app, I used `MultiAppProjectionTransfer` in order to extrapolate these values to the nodes.

```
  [./tau1_to_sub]
    type = MultiAppProjectionTransfer
    fixed_meshes = true
    proj_type = l2
    direction = to_multiapp
    multi_app = sub_app
    source_variable = tau1Aux
    variable = tau1 
  [../]
```
where tau1Aux is a `MONOMIAL` and tau1 is a `LAGRANGE` variables.

Mind you, the mesh between the two apps are exactly identical, in fact, I use `clone_mesh` tag.
is there a better way to do this because this transfer is pretty much holding back the run speed  since I have to do 12 of these transfers?

Thanks in advance!

**Comment by GiudGiud** [comment, marked as answer]
Hello

So this is a performance issue?
Or are you unsatisfied with the granularity of the transfer? I think that projection will not be exactly what you describe here, from Gauss points to nodes

If this is solely a performance issue, `fixed_meshes = true` is already the fastest with ProjectionTransfer
If you need faster, I would recommend working on moving to a `MultiAppCopyTransfer`, our fastest field transfer. For this you will need to have the same type of variable in both applications. I think this will be beneficial for the accuracy of the projections too

Guillaume

**Reply by KhaledNabilSharafeldin** [reply]
It is a performance issue.

I would use the `MultiAppCopyTransfer` if I have these values calculated on the nodes but they are on qp, is there an existing auxkernel or so to interpolate the values from qp to nodes?

Thanks,
Khaled

**Reply by GiudGiud** [reply]
any nodal auxkernel already does this, computes the nodal values based on the qp values. 
So if you have an auxkernel that does the job, it should likely work

in the code if it refers to `_current_elem` instead of qp, typically it doesnt work for nodal variables

**Reply by KhaledNabilSharafeldin** [reply]
It seems like this wont work with material properties defined on qp

it results in the following error:
```
Nodal AuxKernel 'tau1Aux' attempted to reference material property 'mandel_stress'
Consider using an elemental auxiliary variable for 'tau1Aux'.
```

if there is an example AuxKernel that you can refer me to I'd really appreciate it
my auxkernel basically does a double contraction of the `_mandel_stress[_qp].doubleContraction(R);`

**Reply by YaqiWang** [reply]
Do you have to use LAGRANGE type on your target mesh? If not, you can make the two variables in the same type and order and do the copy transfer, which is much faster. If you have to, you still can use a reaction kernel and a body force kernel (to assembly the material property) with tau1Aux being a primal variable. I did not check the code but suppose that the projection transfer can transfer elemental variables to nodal variables and the main cost goes into the setup.

**Reply by GiudGiud** [reply]
There's two things here:
-  saving the material property to a constant monomial loses information, then projecting it using the transfer to a nodal variable will not recover this information, it ll be approximate. If you dont care too much about this, and only want to solve the performance issue, you will want to use the `fixed_meshes` parameter. This will cache most of the work performed in this transfer, making it very fast after the first time it is used
- if you do care about this error, then you will need to implement the nodal version of a materialRealAux. We purposely do not have this because it would not be very well defined at interfaces between blocks. But as long as you make a consistent choice of which block is preferred, or suggest an interpolation method or something, then it will be OK.

**Reply by KhaledNabilSharafeldin** [reply]
Thank you @YaqiWang and @GiudGiud  for your replies!
I took a look at `MaterialPropertyValue` [kernel](https://mooseframework.inl.gov/source/kernels/MaterialPropertyValue.html), and created a version that is compatible with RankTwoTensor material properties using indices. This pretty much does what a reaction or body force kernel would do.
I can share the code here if that would be useful.

Thanks again!
~Khaled
